src/caffe/layers/weighted_softmax_loss_layer.cpp

```diff
@@ -0,0 +1,156 @@
+
+  this->has_ignore_label_ =
+    this->layer_param_.loss_param().has_ignore_label();
+  if (this->has_ignore_label_) {
+    this->ignore_label_ = this->layer_param_.loss_param().ignore_label();
+  }
+  if (!this->layer_param_.loss_param().has_normalization() &&
+      this->layer_param_.loss_param().has_normalize()) {
+    this->normalization_ = this->layer_param_.loss_param().normalize() ?
+                     LossParameter_NormalizationMode_VALID :
+                     LossParameter_NormalizationMode_BATCH_SIZE;
+  } else {
+    this->normalization_ = this->layer_param_.loss_param().normalization();
+  }
+}
+
+template <typename Dtype>
+void WeightedSoftmaxWithLossLayer<Dtype>::Reshape(
+    const vector<Blob<Dtype>*>& bottom, const vector<Blob<Dtype>*>& top) {
+  SoftmaxWithLossLayer<Dtype>::Reshape(bottom, top);
+  tile_layer_->Reshape(tile_bottom_vec_, tile_top_vec_);
+  this->outer_num_ = bottom[0]->count(0, this->softmax_axis_);
+  this->inner_num_ = bottom[0]->count(this->softmax_axis_ + 1);
+  CHECK_EQ(this->outer_num_ * this->inner_num_, bottom[1]->count())
+      << "Number of labels must match number of predictions; "
+      << "e.g., if softmax axis == 1 and prediction shape is (N, C, H, W), "
+      << "label count (number of labels) must be N*H*W, "
+      << "with integer values in {0, 1, ..., C-1}.";
+  CHECK_EQ(bottom[1]->count(), bottom[2]->count())
+      << "Weights and labels must have the same shape";
+  if (top.size() >= 2) {
+    // softmax output
+    top[1]->ReshapeLike(*bottom[0]);
+  }
+}
+
+template <typename Dtype>
+void WeightedSoftmaxWithLossLayer<Dtype>::Forward_cpu(
+    const vector<Blob<Dtype>*>& bottom, const vector<Blob<Dtype>*>& top) {
+  // The forward pass computes the softmax prob values.
+  this->softmax_layer_->Forward(this->softmax_bottom_vec_,
+                                this->softmax_top_vec_);
+  const Dtype* prob_data = this->prob_.cpu_data();
+  const Dtype* label = bottom[1]->cpu_data();
+  const Dtype* weight = bottom[2]->cpu_data();
+  int dim = this->prob_.count() / this->outer_num_;
+  Dtype agg_weight = 0;
+  Dtype loss = 0;
+  for (int i = 0; i < this->outer_num_; ++i) {
+    for (int j = 0; j < this->inner_num_; j++) {
+      const int label_value = static_cast<int>(label[i * this->inner_num_ + j]);
+      const Dtype weight_value = weight[i * this->inner_num_ + j];
+      if (this->has_ignore_label_ && label_value == this->ignore_label_) {
+        continue;
+      }
+      DCHECK_GE(label_value, 0);
+      DCHECK_LT(label_value, this->prob_.shape(this->softmax_axis_));
+      DCHECK_GE(weight_value, 0);  // do not allow negative weights...
+      loss -= weight_value
+        * log(std::max(prob_data[i * dim + label_value * this->inner_num_ + j],
+                           Dtype(FLT_MIN)));
+      agg_weight += weight_value;
+    }
+  }
+  top[0]->mutable_cpu_data()[0] = loss
+    / this->get_normalizer(this->normalization_, agg_weight);
```

Comment:
Have you tried to normalize by count instead of agg_weight? For example, [this](https://github.com/shicai/Weighted_Softmax_Loss/blob/master/weighted_softmax_loss_layer.cpp) weighted softmax loss implementation is normalized by count. Although I think yours makes a bit more sense, otherwise one might need to tune the loss weight a bit.